src/titiler/xarray/titiler/xarray/io.py

```diff
@@ -0,0 +1,230 @@
+from rio_tiler.io.xarray import XarrayReader
+
+
+def xarray_open_dataset(  # noqa: C901
+    src_path: str,
+    group: Optional[Any] = None,
+    decode_times: Optional[bool] = True,
+) -> xarray.Dataset:
+    """Open dataset."""
+    import aiohttp  # noqa
+    import fsspec  # noqa
+    import s3fs  # noqa
+
+    parsed = urlparse(src_path)
+    protocol = parsed.scheme or "file"
+    if protocol not in ["s3", "https", "http", "file"]:
+        raise ValueError(f"Unsupported protocol: {protocol}")
+
+    if any(src_path.lower().endswith(ext) for ext in [".nc", ".nc4"]):
+        xr_engine = "h5netcdf"
+    else:
+        xr_engine = "zarr"
+
+    if protocol == "s3":
+        s3_filesystem = s3fs.S3FileSystem()
+        file_handler = (
+            s3_filesystem.open(src_path)
+            if xr_engine == "h5netcdf"
+            else s3fs.S3Map(root=src_path, s3=s3_filesystem)
+        )
+
+    else:
+        filesystem = fsspec.filesystem(protocol)  # type: ignore
+        file_handler = (
+            filesystem.open(src_path)
+            if xr_engine == "h5netcdf"
+            else filesystem.get_mapper(src_path)
+        )
+
+    # Arguments for xarray.open_dataset
+    # Default args
+    xr_open_args: Dict[str, Any] = {
+        "decode_coords": "all",
+        "decode_times": decode_times,
+    }
+
+    # Argument if we're opening a datatree
+    if group is not None:
+        xr_open_args["group"] = group
+
+    # NetCDF arguments
+    if xr_engine == "h5netcdf":
+        xr_open_args.update(
+            {
+                "engine": "h5netcdf",
+                "lock": False,
+            }
+        )
+
+        ds = xarray.open_dataset(file_handler, **xr_open_args)
+
+    # Fallback to Zarr
+    else:
+        if protocol == "reference":
+            xr_open_args.update(
+                {
+                    "consolidated": False,
+                    "backend_kwargs": {"consolidated": False},
+                }
+            )
+
+        ds = xarray.open_zarr(file_handler, **xr_open_args)
+
+    return ds
+
+
+def _arrange_dims(da: xarray.DataArray) -> xarray.DataArray:
+    """Arrange coordinates and time dimensions.
+
+    An rioxarray.exceptions.InvalidDimensionOrder error is raised if the coordinates are not in the correct order time, y, and x.
+    See: https://github.com/corteva/rioxarray/discussions/674
+
+    We conform to using x and y as the spatial dimension names..
+
+    """
+    if "x" not in da.dims and "y" not in da.dims:
+        try:
+            latitude_var_name = next(
+                x for x in ["lat", "latitude", "LAT", "LATITUDE", "Lat"] if x in da.dims
+            )
+            longitude_var_name = next(
+                x
+                for x in ["lon", "longitude", "LON", "LONGITUDE", "Lon"]
+                if x in da.dims
+            )
+        except StopIteration as e:
+            raise ValueError(f"Couldn't find X/Y dimensions in {da.dims}") from e
+
```

Comment:
Should we add a check here that the `x` and `y` arrays are 1-dimensional?

For example the [UKCP 2.2 km dataset](https://data.ceda.ac.uk/badc/ukcp18/data/land-cpm/uk/2.2km/rcp85/01/tas/mon/v20210615) uses `latitude` and `longitude` as dimension names, but these are 2D fields and reading them with Rasterio/GDAL requires some tweaks https://gist.github.com/j08lue/e792b3c912c33e9191734af7e795b75c

Reply:
Good call to keep the heuristics to a minimum.